platforms/core-runtime/base-services/src/main/java/org/gradle/internal/service/ServiceScopeValidator.java

```diff
@@ -52,24 +52,52 @@ public void whenRegistered(Class<? extends Annotation> annotation, Registration
     }
 
     private void validateScope(Class<?> serviceType) {
-        Class<? extends Scope> serviceScope = scopeOf(serviceType);
+        Class<? extends Scope>[] serviceScopes = scopeOf(serviceType);
 
-        if (serviceScope == null || ServiceScopeValidatorWorkarounds.shouldSuppressValidation(serviceType)) {
+        if (serviceScopes == null) {
             return;
         }
 
-        if (!scope.equals(serviceScope)) {
-            throw new IllegalArgumentException(invalidScopeMessage(serviceType, serviceScope));
+        if (serviceScopes.length == 0) {
+            throw new IllegalArgumentException(String.format("Service '%s' is declared with empty scope list", serviceType.getName()));
+        }
+
+        if (!containsScope(scope, serviceScopes)) {
+            throw new IllegalArgumentException(invalidScopeMessage(serviceType, serviceScopes));
         }
     }
 
-    private String invalidScopeMessage(Class<?> serviceType, Class<? extends Scope> actualScope) {
-        return String.format("Service '%s' was declared in scope '%s' but registered in scope '%s'",
-            serviceType.getName(), actualScope.getSimpleName(), scope.getSimpleName());
+    private static boolean containsScope(Class<? extends Scope> expectedScope, Class<? extends Scope>[] actualScopes) {
```

Comment:
💅 omg, Java! Let's refactor this into something like `org.gradle.internal.util.ArrayExtensions#contains`